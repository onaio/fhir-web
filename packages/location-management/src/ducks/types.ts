import { Geometry } from 'geojson';
import { ParsedHierarchyNode, TreeNode } from './locationHierarchy/types';

/** The shape of a jurisdiction received from the OpenSRP API */
export interface OpenSRPJurisdiction {
  id: string;
  geometry?: Geometry;
  properties: {
    code?: string;
    geographicLevel: number;
    name: string;
    parentId?: string;
    status: string;
    version: string | number;
  };
  serverVersion: number;
  type: 'Feature';
}

export interface LocationTree {
  keys: React.Key[];
  node: ParsedHierarchyNode;
}

export type LocationTreeState = LocationTree;

/** Object containing known API endpoints by name */
export interface APIEndpoints {
  [key: string]: string;
}

/** Convenient type for either an array of jurisdictions or of tree nodes */
export type TreeNodeType = OpenSRPJurisdiction | TreeNode;
