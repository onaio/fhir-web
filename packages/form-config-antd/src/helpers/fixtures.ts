/* eslint-disable @typescript-eslint/camelcase */
export const manifestRelease4 = {
  identifier: '1.0.30',
  json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
  appId: 'org.smartregister.foo',
  appVersion: '1.2.18',
  createdAt: '2020-06-09T14:40:04.189+02:00',
  updatedAt: '2020-06-09T14:40:04.189+02:00',
};

export const fixManifestReleases = [
  {
    identifier: '1.0.11',
    json: '{"forms_version":"1.0.11","identifiers":["blood_screening.json"}',
    appId: 'org.smartregister.reveal',
    appVersion: '1.2.14',
    createdAt: '2020-06-08T15:00:38.413+02:00',
    updatedAt: '2020-06-08T15:00:38.413+02:00',
  },
  {
    identifier: '1.0.12',
    json: '{"forms_version":"1.0.12","identifiers":["blood_screening.json"}',
    appId: 'org.smartregister.reveal',
    appVersion: '1.2.14',
    createdAt: '2020-06-08T15:03:00.950+02:00',
    updatedAt: '2020-06-08T15:03:00.950+02:00',
  },
  {
    identifier: '1.0.16',
    json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
    appId: 'org.smartregister.reveal',
    appVersion: '1.2.16',
    createdAt: '2020-06-09T14:40:04.189+02:00',
    updatedAt: '2020-06-09T14:40:04.189+02:00',
  },
];

export const manifestFile1 = {
  createdAt: 'Jun 19, 2020, 12:31:37 PM',
  form_relation: '',
  id: '52',
  identifier: 'test-form-1.json',
  isDraft: false,
  isJsonValidator: false,
  jursdiction: '',
  label: 'test form',
  module: '',
  version: '1.0.26',
};

export const manifestFile2 = {
  createdAt: 'Jun 19, 2020, 4:23:22 PM',
  form_relation: '',
  id: '53',
  identifier: 'reveal-test-file.json',
  isDraft: false,
  isJsonValidator: false,
  jursdiction: '',
  label: 'test publish',
  module: '',
  version: '1.0.27',
};

export const manifestFile3 = {
  createdAt: 'Jun 19, 2020, 6:23:22 PM',
  form_relation: '',
  id: '54',
  identifier: 'foo.json',
  isDraft: false,
  isJsonValidator: false,
  jursdiction: '',
  label: 'foo',
  module: 'baz',
  version: '1.0.28',
};

export const draftFile1 = {
  ...manifestFile1,
  isDraft: true,
  isJsonValidator: false,
};

export const draftFile2 = {
  ...manifestFile2,
  isDraft: true,
  isJsonValidator: false,
};

export const draftFile3 = {
  ...manifestFile3,
  isDraft: true,
  isJsonValidator: false,
};

export const fixManifestFiles = [manifestFile1, manifestFile2];

export const FixManifestDraftFiles = [
  {
    ...fixManifestFiles[0],
    isDraft: false,
  },
  {
    ...fixManifestFiles[1],
    isDraft: false,
  },
];

export const downloadFile = {
  clientForm: {
    json: JSON.stringify(fixManifestFiles[0]),
  },
};
