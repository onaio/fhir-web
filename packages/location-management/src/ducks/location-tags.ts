import {
  fetchActionCreatorFactory,
  removeActionCreatorFactory,
  setTotalRecordsFactory,
  reducerFactory,
  getItemsByIdFactory,
  getItemsArrayFactory,
  getItemByIdFactory,
  getTotalRecordsFactory,
} from '@opensrp/reducer-factory';

/** interface for location group **/
export interface LocationTag {
  id: number;
  active: boolean;
  name: string;
  description: string;
}

/** reducer name for the Item module */
export const reducerName = 'location-tags';

/** Action types */
const customfetchedActionType = 'location-tags/LOCATION_TAGS_FETCHED';
const customRemoveActionType = 'location-tags/REMOVE_LOCATION_TAGS';
const customSetTotalRecordsActionType = 'location-tags/SET_TOTAL_LOCATION_TAGS';

/** Item Reducer */
const reducer = reducerFactory<LocationTag>(
  reducerName,
  customfetchedActionType,
  customRemoveActionType,
  customSetTotalRecordsActionType
);

// action
/** actionCreator returns action to to add Item records to store */
export const fetchLocationTags = fetchActionCreatorFactory<LocationTag>(reducerName, 'id');
export const removeLocationTags = removeActionCreatorFactory(reducerName);
export const setTotalLocationtags = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationTagsById = getItemsByIdFactory<LocationTag>(reducerName);
export const getLocationTagById = getItemByIdFactory<LocationTag>(reducerName);
export const getLocationTagsArray = getItemsArrayFactory<LocationTag>(reducerName);
export const getTotalLocationTags = getTotalRecordsFactory(reducerName);

export default reducer;
