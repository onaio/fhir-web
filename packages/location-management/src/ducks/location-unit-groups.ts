import {
  fetchActionCreatorFactory,
  removeActionCreatorFactory,
  setTotalRecordsFactory,
  reducerFactory,
  getItemsByIdFactory,
  getItemsArrayFactory,
  getItemByIdFactory,
  getTotalRecordsFactory,
} from '@opensrp/reducer-factory';

/** interface for location Unit Group **/
export interface LocationUnitGroup {
  id: number;
  active: boolean;
  name: string;
  description: string;
}

export interface LocationUnitGroupPayloadPOST {
  active: boolean;
  name: string;
  description: string;
}

export interface LocationUnitGroupPayloadPUT extends LocationUnitGroupPayloadPOST {
  id: string;
}

/** reducer name for the Item module */
export const reducerName = 'location-unit-groups';

/** Action types */
const customfetchedActionType = 'location-unit-groups/LOCATION_UNIT_GROUPS_FETCHED';
const customRemoveActionType = 'location-unit-groups/REMOVE_LOCATION_UNIT_GROUPS';
const customSetTotalRecordsActionType = 'location-unit-groups/SET_TOTAL_LOCATION_UNIT_GROUPS';

/** Item Reducer */
export const reducer = reducerFactory<LocationUnitGroup>(
  reducerName,
  customfetchedActionType,
  customRemoveActionType,
  customSetTotalRecordsActionType
);

// action
/** actionCreator returns action to to add Item records to store */
export const fetchLocationUnitGroups = fetchActionCreatorFactory<LocationUnitGroup>(
  reducerName,
  'id'
);
export const removeLocationUnitGroups = removeActionCreatorFactory(reducerName);
export const setTotalLocationunitgroups = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationUnitGroupsById = getItemsByIdFactory<LocationUnitGroup>(reducerName);
export const getLocationUnitGroupById = getItemByIdFactory<LocationUnitGroup>(reducerName);
export const getLocationUnitGroupsArray = getItemsArrayFactory<LocationUnitGroup>(reducerName);
export const getTotalLocationUnitGroups = getTotalRecordsFactory(reducerName);

export default reducer;
