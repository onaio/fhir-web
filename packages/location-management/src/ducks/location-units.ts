import {
  fetchActionCreatorFactory,
  removeActionCreatorFactory,
  setTotalRecordsFactory,
  reducerFactory,
  getItemsByIdFactory,
  getItemsArrayFactory,
  getItemByIdFactory,
  getTotalRecordsFactory,
} from '@opensrp/reducer-factory';
import { Geometry } from 'geojson';
import { values } from 'lodash';
import { Store } from 'redux';

/** Enum representing the possible location unit status types */
export enum LocationUnitStatus {
  ACTIVE = 'Active',
  INACTIVE = 'InActive',
}

export enum LocationUnitSyncStatus {
  SYNCED = 'Synced',
  NOTSYNCED = 'NotSynced',
}

/** interface for LocationUnit.properties */
export interface Properties {
  name: string;
  parentId: string;
  status: LocationUnitStatus;
  geographicLevel?: number;
  username?: string;
  version?: number;
  name_en?: string;
  externalId?: string;
}

/** location unit tag interface */
export interface LocationUnitTag {
  id: number;
  name: string;
}

/** location interface */
export interface LocationUnit {
  id: string | number;
  properties: Properties;
  syncStatus: LocationUnitSyncStatus;
  type: string;
  locationTags?: LocationUnitTag[];
  geometry?: Geometry;
}

/** interface for the POST payload */
export interface LocationUnitPayloadPOST {
  properties?: Properties;
  syncStatus?: LocationUnitSyncStatus;
  type: string;
  locationTags?: LocationUnitTag[];
  geometry?: Geometry;
  textEntry?: string[];
  // we will remove this id as it should be auto generated on server
  id?: string | number;
}

/** interface for the PUT payload */
export interface LocationUnitPayloadPUT extends LocationUnitPayloadPOST {
  id: string | number;
}

/** reducer name for the Item module */
export const reducerName = 'location-units';

/** Action types */
const customfetchedActionType = 'location-units/LOCATION_UNITS_FETCHED';
const customRemoveActionType = 'location-units/REMOVE_LOCATION_UNITS';
const customSetTotalRecordsActionType = 'location-units/SET_TOTAL_LOCATION_UNITS';

/** Item Reducer */
const reducer = reducerFactory<LocationUnit>(
  reducerName,
  customfetchedActionType,
  customRemoveActionType,
  customSetTotalRecordsActionType
);

// action
/** actionCreator returns action to to add Item records to store */
export const fetchLocationUnits = fetchActionCreatorFactory<LocationUnit>(reducerName, 'id');
export const removeLocationUnits = removeActionCreatorFactory(reducerName);
export const setTotalLocationUnits = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationUnitsById = getItemsByIdFactory<LocationUnit>(reducerName);
export const getLocationUnitById = getItemByIdFactory<LocationUnit>(reducerName);
export const getLocationUnitsArray = getItemsArrayFactory<LocationUnit>(reducerName);
export const getTotalLocationUnits = getTotalRecordsFactory(reducerName);

export const LocationUnitsArray = (state: Partial<Store>): LocationUnit[] =>
  values(getLocationUnitsById(state) || {});

export default reducer;
