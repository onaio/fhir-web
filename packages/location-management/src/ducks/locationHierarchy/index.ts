/** store location hierarchy information
 * - should be able to add a tree model
 * - modify the tree model in a manner such that shallow comparison would yield false
 * - provide the selectors
 */

import { Dictionary } from '@onaio/utils';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { RawOpenSRPHierarchy, TreeNode } from './types';
import { generateJurisdictionTree } from './utils';
import { createSelector } from 'reselect';
import { values } from 'lodash';

/** reducer name for hierarchy reducer */
export const hierarchyReducerName = 'locationHierarchy';

// **************************** actions *****************************

/** action to add a tree to store */
export const TREE_FETCHED = 'opensrp/locations/hierarchy/TREE_FETCHED';

/** remove trees */
export const DEFOREST = 'opensrp/locations/hierarchy/DEFOREST';

/** describes action that adds a tree to store */
export interface FetchedTreeAction extends AnyAction {
  type: typeof TREE_FETCHED;
  treeByRootId: Dictionary<TreeNode>;
}

/** describes action to remove all trees */
export interface DeforestAction extends AnyAction {
  type: typeof DEFOREST;
  treeByRootId: {};
}

/** combined full action types | its a union */
export type TreeActionTypes = FetchedTreeAction | DeforestAction | AnyAction;

// **************************** action creators ****************************

/** action creator when adding a tree to store
 *
 * @param apiResponse - the raw hierarchy as received from opensrp
 * @param treeId - the treeId to use while saving to the store
 */
export function fetchTree(
  apiResponse: RawOpenSRPHierarchy,
  treeId: string | null = null
): FetchedTreeAction {
  const tree = generateJurisdictionTree(apiResponse);
  return {
    treeByRootId: {
      [treeId ? treeId : tree.model.id]: tree,
    },
    type: TREE_FETCHED,
  };
}

/** clear all the trees */
export function deforest(): DeforestAction {
  return {
    treeByRootId: {},
    type: DEFOREST,
  };
}

// **************************** reducer ****************************

/** The store's slice state
 */
export interface TreeState {
  treeByRootId: Dictionary<TreeNode> | {};
}

/** Create an immutable tree state */
export type ImmutableTreeState = TreeState & SeamlessImmutable.ImmutableObject<TreeState>;

/** starting state */
export const initialState: ImmutableTreeState = SeamlessImmutable({
  treeByRootId: {},
});

// the reducer function
/**
 *
 * @param {Dictionary} state - the store
 * @param {AnyAction} action - the redux action
 * @returns {object} - updated state
 */
export function hierarchyReducer(
  state: ImmutableTreeState | Dictionary = initialState,
  action: TreeActionTypes
) {
  switch (action.type) {
    case TREE_FETCHED:
      return {
        ...state,
        treeByRootId: { ...state.treeByRootId, ...action.treeByRootId },
      };
    case DEFOREST:
      return {
        ...state,
        treeByRootId: action.treeByRootId,
      };
    default:
      return state;
  }
}

/** prop filters to customize selector queries */
export interface Filters {
  rootJurisdictionId?: string[] /** specify which tree to act upon, undefined allows all */;
  geoLevel?: number /** get locations at the specified level */;
  searchQuery?: string /** search query to filter nodes against, by name and id */;
}

/** retrieve the rootJurisdiction value
 *
 * @param _ - the store
 * @param props -  the filterProps
 */
export const getRootJurisdictionId = (_: Partial<Store>, props: Filters) =>
  props.rootJurisdictionId;

/** retrieve the geoLevel value
 *
 * @param _ - the store
 * @param props -  the filterProps
 */
export const getGeoLevel = (_: Partial<Store>, props: Filters) => props.geoLevel;

/** retrieve the geoLevel value
 *
 * @param _ - the store
 * @param props -  the filterProps
 */
export const getSearchQuery = (_: Partial<Store>, props: Filters) => props.searchQuery;

/** gets all trees key'd by the rootNodes id
 *
 * @param state - the store
 * @param _ -  the filterProps
 */
export const getTreesByRootId = (state: Partial<Store>, _: Filters): Dictionary<TreeNode> =>
  (state as Dictionary)[hierarchyReducerName].treeByRootId;

/** factory that returns a selector to retrieve the tree(s) using their rootNode's ids */
export const getTreesByIds = () =>
  createSelector(getTreesByRootId, getRootJurisdictionId, (treesById, rootIds): TreeNode[] => {
    const allTreesArray = values(treesById);
    if (rootIds === undefined) {
      return allTreesArray;
    }
    const treesOfInterest: TreeNode[] = [];
    rootIds.forEach((rootId) => {
      const treeOfInterest = treesById[rootId] as TreeNode | undefined;
      if (treeOfInterest) {
        treesOfInterest.push(treeOfInterest);
      }
    });
    return treesOfInterest;
  });

const treeSelectors = getTreesByIds();

/** factory that returns a selector to retrieve the tree(s) using their rootNode's ids
 * the selector will return all nodes if geographic level is undefined
 */
export const getLocationsByLevel = () =>
  createSelector(treeSelectors, getGeoLevel, (trees, geoLevel): TreeNode[] => {
    let locationsOfInterest: TreeNode[] = [];
    trees.forEach((tree) => {
      locationsOfInterest = [
        ...locationsOfInterest,
        ...tree.all((node) => {
          if (geoLevel === undefined) {
            return true;
          }
          return node.model.node.attributes.geographicLevel === geoLevel;
        }),
      ];
    });
    return locationsOfInterest;
  });

const geoLevelSelector = getLocationsByLevel();

/** factory that returns a selector that can be used to filter the nodes by either their
 * labels or ids
 */
export const getLocationsByNameAndId = () =>
  createSelector(geoLevelSelector, getSearchQuery, (nodes, searchQuery): TreeNode[] => {
    if (searchQuery === undefined) {
      return nodes;
    }
    const matchesName = (node: TreeNode) =>
      node.model.label.toLowerCase().includes(searchQuery.toLowerCase());
    const matchesId = (node: TreeNode) => node.model.id === searchQuery;
    return nodes.filter((node) => {
      return matchesName(node) || matchesId(node);
    });
  });
