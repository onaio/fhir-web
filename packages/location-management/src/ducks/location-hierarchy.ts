/** store location hierarchy information
 * - should be able to add a tree model
 * - modify the tree model in a manner such that shallow comparison would yield false
 * - provide the selectors
 */

import { Dictionary } from '@onaio/utils';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { TreeNode } from './locationHierarchy/types';
import { LocationTreeState } from './types';

/** reducer name for hierarchy reducer */
export const reducerName = 'location-hierarchy';

// **************************** actions *****************************

/** action to add a tree to store */
export const TREE_FETCHED = 'location-hierarchy/TREE_FETCHED';
export const FETCH_SINGLE_LOCATION = 'location-hierarchy/FETCH_SINGLE_LOCATION';
export const SET_LOCATION_TREE_STATE = 'location-hierarchy/SET_LOCATION_TREE_STATE';

/** describes action that adds a hierarchy tree to store */
export interface FetchedTreeAction extends AnyAction {
  type: typeof TREE_FETCHED;
  hierarchyObject: TreeNode;
}

/** describes action that saves a hierarchy tree to store */
export interface SetLocationTreeStateAction extends AnyAction {
  type: typeof SET_LOCATION_TREE_STATE;
  hierarchyObject: LocationTreeState;
}

/** combined full action types | its a union */
export type TreeActionTypes = FetchedTreeAction | SetLocationTreeStateAction | AnyAction;

// **************************** action creators ****************************

/** action creator when adding a tree to store
 *
 * @param {TreeNode} hierarchy - the raw hierarchy as received from opensrp
 * @returns {object} - action object
 */
export function fetchAllHierarchies(hierarchy: TreeNode): FetchedTreeAction {
  return {
    hierarchyObject: {
      ...hierarchy,
    } as TreeNode,
    type: TREE_FETCHED,
  };
}

/** action creator when adding a tree to store
 *
 * @param {LocationTreeState} hierarchy - the tree state to be expanded
 * @returns {object} - action object
 */
export function setLocationTreeState(hierarchy: LocationTreeState): SetLocationTreeStateAction {
  return {
    hierarchyObject: {
      ...hierarchy,
    } as LocationTreeState,
    type: SET_LOCATION_TREE_STATE,
  };
}

// **************************** medusa ****************************

/** The store's slice state */
export interface TreeState {
  hierarchyArray: TreeNode[];
  currentParentChildren: TreeNode[];
  locationTreeState: LocationTreeState;
}

/** Create an immutable tree state */
export type ImmutableTreeState = TreeState & SeamlessImmutable.ImmutableObject<TreeState>;

/** starting state */
export const initialState: ImmutableTreeState | Dictionary = SeamlessImmutable({
  hierarchyArray: [],
  currentParentChildren: [],
  locationObject: {},
  locationTreeState: {},
});

// the reducer function
/**
 *
 * @param {Dictionary} state - the store
 * @param {AnyAction} action - the redux action
 * @returns {object} - updated state
 */
export function reducer(
  state: ImmutableTreeState | Dictionary = initialState,
  action: TreeActionTypes
) {
  switch (action.type) {
    case TREE_FETCHED:
      return {
        ...state,
        hierarchyArray: [...state.hierarchyArray, action.hierarchyObject],
      };

    case SET_LOCATION_TREE_STATE:
      return {
        ...state,
        locationTreeState: action.hierarchyObject,
      };

    default:
      return state;
  }
}

/** gets array of all hierarchies
 *
 * @param {Store} state - the store
 * @returns {object} - returns item from location-hierarchy reducer
 */
export const getAllHierarchiesArray = (state: Partial<Store>): Dictionary<TreeNode> =>
  (state as Dictionary)[reducerName].hierarchyArray;

export const getLocationTreeState = (state: Partial<Store>): Dictionary<LocationTreeState> =>
  (state as Dictionary)[reducerName].locationTreeState;
