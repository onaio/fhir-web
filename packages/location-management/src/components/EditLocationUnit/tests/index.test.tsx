import { mount, shallow } from 'enzyme';
import { EditLocationUnit } from '..';
import React from 'react';
import { store } from '@opensrp/store';
import { createBrowserHistory } from 'history';
import { RouteComponentProps, Router } from 'react-router';
import { Provider } from 'react-redux';
import { Helmet } from 'react-helmet';
import { removeLocationUnits } from '../../../ducks/location-units';
import { authenticateUser } from '@onaio/session-reducer';
import { location1 } from '../../LocationForm/tests/fixtures';
import { act } from 'react-dom/test-utils';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('jest-fetch-mock');

const history = createBrowserHistory();

const path = '/locations/edit';
const locationId = location1.id;
const locationProps = {
  history,
  location: {
    hash: '',
    pathname: `${'/locations/edit/'}`,
    search: '',
    state: {},
  },
  match: {
    isExact: true,
    params: { id: locationId },
    path: `${path}`,
    url: `${path}/${locationId}`,
  },
};

describe('EditLocationUnit', () => {
  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        // eslint-disable-next-line @typescript-eslint/camelcase
        { api_token: 'hunter2', oAuth2Data: { access_token: 'sometoken', state: 'abcde' } }
      )
    );
  });

  afterEach(() => {
    fetch.resetMocks();
    store.dispatch(removeLocationUnits());
  });

  it('renders without crashing', async () => {
    fetch.mockResponse(JSON.stringify([]));
    shallow(
      <Provider store={store}>
        <Router history={history}>
          <EditLocationUnit {...locationProps} />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
    });
  });

  it('renders correctly when location is jurisdiction', async () => {
    fetch.once(JSON.stringify(location1));
    fetch.once(JSON.stringify(null));
    fetch.mockResponse(JSON.stringify([]));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditLocationUnit {...locationProps} />
        </Router>
      </Provider>
    );

    // loading page
    expect(wrapper.text()).toMatchInlineSnapshot(`""`);

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    const helmet = Helmet.peek();
    expect(helmet.title).toEqual('Edit > Kenya');

    // rendered page including title
    expect(wrapper.find('h5').text()).toMatchInlineSnapshot(`"Edit > Kenya"`);

    expect(wrapper.find('LocationForm').text()).toMatchSnapshot('form rendered');

    expect(fetch.mock.calls[0]).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/rest/location/b652b2f4-a95d-489b-9e28-4629746db96a?return_geometry=true&is_jurisdiction=true',
      {
        headers: {
          accept: 'application/json',
          authorization: 'Bearer sometoken',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'GET',
      },
    ]);
    expect(fetch.mock.calls[1]).toEqual([
      'https://opensrp-stage.smartregister.org/opensrp/rest/location/b652b2f4-a95d-489b-9e28-4629746db96a?return_geometry=true&is_jurisdiction=false',
      {
        headers: {
          accept: 'application/json',
          authorization: 'Bearer sometoken',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'GET',
      },
    ]);

    // check isJurisdiction status passed to form
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect((wrapper.find('LocationForm').props() as any).initialValues.isJurisdiction).toBeTruthy();
  });

  it('renders correctly when location is structure', async () => {
    fetch.once(JSON.stringify(null));
    fetch.once(JSON.stringify(location1));
    fetch.mockResponse(JSON.stringify([]));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditLocationUnit {...locationProps} />
        </Router>
      </Provider>
    );

    // loading page
    expect(wrapper.text()).toMatchInlineSnapshot(`""`);

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // check isJurisdiction status passed to form
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    expect((wrapper.find('LocationForm').props() as any).initialValues.isJurisdiction).toBeFalsy();
  });

  it('renders errorPage correctly', async () => {
    const errorMessage = 'An error happened';
    fetch.mockReject(new Error(errorMessage));

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditLocationUnit {...locationProps} />
        </Router>
      </Provider>
    );

    // loading page
    expect(wrapper.text()).toMatchInlineSnapshot(`""`);

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.text()).toMatchInlineSnapshot(`"ErrorAn error happenedGo BackBack Home"`);
  });

  it('renders resource404 when location is not found', async () => {
    fetch.once(JSON.stringify(null));
    fetch.once(JSON.stringify(location1));
    fetch.mockResponse(JSON.stringify([]));

    const props = {
      ...locationProps,
      match: {
        ...locationProps.match,
        params: { id: 'unknown' },
      },
    };

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditLocationUnit {...props} />
        </Router>
      </Provider>
    );

    // loading page
    expect(wrapper.text()).toMatchInlineSnapshot(`""`);

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // resource not found
    expect(wrapper.text()).toMatchInlineSnapshot(
      `"404Sorry, the resource you requested for, does not existGo BackBack Home"`
    );
  });

  it('cancel url is used if passed', async () => {
    fetch.once(JSON.stringify(null));
    fetch.once(JSON.stringify(location1));
    fetch.mockResponse(JSON.stringify([]));
    const cancelURL = '/canceledURL';

    const props = {
      ...locationProps,
      match: {
        ...locationProps.match,
        params: { id: location1.id },
      },
      cancelURLGenerator: () => cancelURL,
    };

    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <EditLocationUnit {...props} />
        </Router>
      </Provider>
    );

    await act(async () => {
      await new Promise((resolve) => setImmediate(resolve));
      wrapper.update();
    });

    // simulate click on cancel button
    wrapper.find('button#location-form-cancel-button').simulate('click');
    wrapper.update();

    // check url
    expect(
      (wrapper.find('Router').props() as RouteComponentProps).history.location.pathname
    ).toEqual(cancelURL);
  });
});
