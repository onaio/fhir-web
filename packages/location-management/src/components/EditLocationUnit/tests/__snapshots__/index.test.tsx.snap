// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLocationUnit renders correctly when location is jurisdiction: form rendered 1`] = `"InstanceIdusernameParentNameStatusActiveInactiveLocation categoryService pointJurisdictionTypeTypeSelect the service point typeExternal IDGeometryUnit group23Â SaveCancel"`;
